File: cdc/src/main/java/org/neo4j/cdc/client/pattern/Visitors.java
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.neo4j.cdc.client.pattern;

import java.util.*;
import java.util.stream.Collectors;
import org.antlr.v4.runtime.*;

class Visitors {

    static List<Pattern> parse(String source) {
        var input = CharStreams.fromString(source);
        var lexer = new PatternLexer(input);
        var tokens = new CommonTokenStream(lexer);
        var parser = new PatternParser(tokens);

        var errors = new ArrayList<String>();
        parser.addErrorListener(new BaseErrorListener() {
            @Override
            public void syntaxError(
                    Recognizer<?, ?> recognizer,
                    Object offendingSymbol,
                    int line,
                    int charPositionInLine,
                    String msg,
                    RecognitionException e) {
                errors.add(String.format("line %d:%d, %s", line, charPositionInLine, msg));
            }
        });

        var patternList = parser.patternList();

        if (!errors.isEmpty()) {
            throw new PatternException(String.join(", ", errors));
        }

        return patternList.pattern().stream()
                .map(PatternVisitor.INSTANCE::visitPattern)
                .collect(Collectors.toList());
    }

    private static class PatternVisitor extends PatternParserBaseVisitor<Pattern> {
        public static final PatternVisitor INSTANCE = new PatternVisitor();

        @Override
        public Pattern visitPattern(PatternParser.PatternContext ctx) {
            if (ctx.relationshipPattern() != null) {
                var relPattern = ctx.relationshipPattern();
                var bidirectional = (relPattern.leftArrow() != null && relPattern.rightArrow() != null)
                        || (relPattern.leftArrow() == null && relPattern.rightArrow() == null);
                var keyFilters = new HashMap<String, Object>();
                var includeProperties = new HashSet<String>();
                var excludeProperties = new HashSet<String>();

                if (relPattern.properties() != null && relPattern.properties().propertySelector() != null) {
                    extractPropertySelectors(
                            relPattern.properties().propertySelector().keyFilterOrPropSelector(),
                            keyFilters,
                            includeProperties,
                            excludeProperties);
                }

                var startNode =
                        visitNodePattern(ctx.nodePattern((bidirectional || relPattern.rightArrow() != null) ? 0 : 1));
                if (!startNode.getExcludeProperties().isEmpty()
                        || !startNode.getIncludeProperties().isEmpty()) {
                    throw new PatternException(
                            "property selectors are not allowed in node part of relationship patterns");
                }

                var endNode =
                        visitNodePattern(ctx.nodePattern((bidirectional || relPattern.rightArrow() != null) ? 1 : 0));
                if (!endNode.getExcludeProperties().isEmpty()
                        || !endNode.getIncludeProperties().isEmpty()) {
                    throw new PatternException(
                            "property selectors are not allowed in node part of relationship patterns");
                }

                return new RelationshipPattern(
                        LabelOrRelTypeVisitor.INSTANCE.visitLabelOrRelType(relPattern.labelOrRelType()),
                        startNode,
                        endNode,
                        bidirectional,
                        keyFilters,
                        includeProperties,
                        excludeProperties);
            }

            return visitNodePattern(ctx.nodePattern(0));
        }

        @Override
        public NodePattern visitNodePattern(PatternParser.NodePatternContext ctx) {
            var labels = NodeLabelsVisitor.INSTANCE.visitNodeLabels(ctx.nodeLabels());
            var keyFilters = new HashMap<String, Object>();
            var includeProperties = new HashSet<String>();
            var excludeProperties = new HashSet<String>();

            if (ctx.properties() != null && ctx.properties().propertySelector() != null) {
                extractPropertySelectors(
                        ctx.properties().propertySelector().keyFilterOrPropSelector(),
                        keyFilters,
                        includeProperties,
                        excludeProperties);
            }

            return new NodePattern(labels, keyFilters, includeProperties, excludeProperties);
        }

        private void extractPropertySelectors(
                List<PatternParser.KeyFilterOrPropSelectorContext> selectors,
                Map<String, Object> keyFilters,
                Set<String> includeProperties,
                Set<String> excludeProperties) {
            selectors.forEach(child -> {
                if (child.keyFilter() != null) {
                    var propName = PropertyKeyNameVisitor.INSTANCE.visitPropertyKeyName(
                            child.keyFilter().propertyKeyName());
                    var value = ExpressionVisitor.INSTANCE.visitExpression(
                            child.keyFilter().expression());

                    keyFilters.put(propName, value);
                } else if (child.propSelector() != null) {
                    if (child.propSelector().TIMES() != null) {
                        includeProperties.add("*");
                    } else if (child.propSelector().MINUS() != null) {
                        excludeProperties.add(PropertyKeyNameVisitor.INSTANCE.visitPropertyKeyName(
                                child.propSelector().propertyKeyName()));
                    } else {
                        includeProperties.add(PropertyKeyNameVisitor.INSTANCE.visitPropertyKeyName(
                                child.propSelector().propertyKeyName()));
                    }
                }
            });
        }
    }

    private static class NodeLabelsVisitor extends PatternParserBaseVisitor<Set<String>> {
        public static final NodeLabelsVisitor INSTANCE = new NodeLabelsVisitor();

        public Set<String> visitNodeLabels(PatternParser.NodeLabelsContext ctx) {
            if (ctx == null || ctx.labelOrRelType() == null) {
                return Set.of();
            }

            return ctx.labelOrRelType().stream()
                    .map(LabelOrRelTypeVisitor.INSTANCE::visitLabelOrRelType)
                    .collect(Collectors.toSet());
        }
    }

    private static class LabelOrRelTypeVisitor extends PatternParserBaseVisitor<String> {
        public static final LabelOrRelTypeVisitor INSTANCE = new LabelOrRelTypeVisitor();

        public String visitLabelOrRelType(PatternParser.LabelOrRelTypeContext ctx) {
            if (ctx == null) {
                return null;
            }

            return SymbolicNameStringVisitor.INSTANCE.visitSymbolicNameString(ctx.symbolicNameString());
        }
    }

    private static class PropertyKeyNameVisitor extends PatternParserBaseVisitor<String> {
        public static final PropertyKeyNameVisitor INSTANCE = new PropertyKeyNameVisitor();

        public String visitPropertyKeyName(PatternParser.PropertyKeyNameContext ctx) {
            return SymbolicNameStringVisitor.INSTANCE.visitSymbolicNameString(ctx.symbolicNameString());
        }
    }

    private static class SymbolicNameStringVisitor extends PatternParserBaseVisitor<String> {
        public static final SymbolicNameStringVisitor INSTANCE = new SymbolicNameStringVisitor();

        public String visitSymbolicNameString(PatternParser.SymbolicNameStringContext ctx) {
            return super.visitSymbolicNameString(ctx);
        }

        public String visitEscapedSymbolicNameString(PatternParser.EscapedSymbolicNameStringContext ctx) {
            return ctx.getText();
        }

        public String visitUnescapedSymbolicNameString(PatternParser.UnescapedSymbolicNameStringContext ctx) {
            return ctx.getText();
        }
    }

    private static class ExpressionVisitor extends PatternParserBaseVisitor<Object> {
        public static final ExpressionVisitor INSTANCE = new ExpressionVisitor();

        public Object visitExpression(PatternParser.ExpressionContext ctx) {
            if (ctx.literal() != null) {
                return ctx.literal().accept(LiteralVisitor.INSTANCE);
            } else {
                throw new PatternException(String.format("unsupported expression: '%s'", ctx));
            }
        }
    }

    private static class LiteralVisitor extends PatternParserBaseVisitor<Object> {
        private static final LiteralVisitor INSTANCE = new LiteralVisitor();

        public Object visitKeywordLiteral(PatternParser.KeywordLiteralContext ctx) {
            if (ctx.INF() != null || ctx.INFINITY() != null) {
                return Double.POSITIVE_INFINITY;
            } else if (ctx.NAN() != null) {
                return Double.NaN;
            } else {
                throw new PatternException(String.format("unsupported keyword literal: '%s'", ctx));
            }
        }

        public Object visitBooleanLiteral(PatternParser.BooleanLiteralContext ctx) {
            return ctx.TRUE() != null;
        }

        public Object visitOtherLiteral(PatternParser.OtherLiteralContext ctx) {
            if (ctx.mapLiteral() != null) {
                var result = new HashMap<String, Object>();
                var mapProps = ctx.mapLiteral().propertyKeyName();
                var mapExps = ctx.mapLiteral().expression();
                for (int i = 0; i < mapProps.size(); i++) {
                    result.put(
                            SymbolicNameStringVisitor.INSTANCE.visit(mapProps.get(i)),
                            ExpressionVisitor.INSTANCE.visitExpression(mapExps.get(i)));
                }
                return result;
            } else if (ctx.listLiteral() != null) {
                return ctx.listLiteral().expression().stream()
                        .map(ExpressionVisitor.INSTANCE::visitExpression)
                        .collect(Collectors.toList());
            } else {
                throw new PatternException(String.format("unsupported literal: '%s'", ctx));
            }
        }

        public Object visitStringsLiteral(PatternParser.StringsLiteralContext ctx) {
            var text = ctx.getText();
            return text.substring(1, text.length() - 1);
        }

        public Object visitNumericLiteral(PatternParser.NumericLiteralContext ctx) {
            var number = ctx.numberLiteral();
            var multiplier = number.MINUS() != null ? -1 : 1;

            if (number.DECIMAL_DOUBLE() != null) {
                return Double.parseDouble(number.DECIMAL_DOUBLE().getText()) * multiplier;
            } else if (number.UNSIGNED_DECIMAL_INTEGER() != null) {
                return Long.parseLong(number.UNSIGNED_DECIMAL_INTEGER().getText(), 10) * multiplier;
            } else if (number.UNSIGNED_HEX_INTEGER() != null) {
                return Long.parseLong(number.UNSIGNED_HEX_INTEGER().getText().substring(2), 16) * multiplier;
            } else if (number.UNSIGNED_OCTAL_INTEGER() != null) {
                return Long.parseLong(number.UNSIGNED_OCTAL_INTEGER().getText().substring(2), 8) * multiplier;
            } else {
                throw new PatternException(String.format("unexpected number literal '%s'", ctx));
            }
        }
    }
}

File: sink/src/main/kotlin/org/neo4j/connectors/kafka/sink/strategy/pattern/Visitors.kt
/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [https://neo4j.com]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.neo4j.connectors.kafka.sink.strategy.pattern

import org.antlr.v4.runtime.BaseErrorListener
import org.antlr.v4.runtime.CharStreams
import org.antlr.v4.runtime.CommonTokenStream
import org.antlr.v4.runtime.RecognitionException
import org.antlr.v4.runtime.Recognizer
import org.neo4j.connectors.kafka.sink.strategy.pattern.PatternParser.PropertyKeyNameOrAliasContext

internal object Visitors {
  fun parse(source: String?): Pattern {
    val input = CharStreams.fromString(source)
    val lexer = PatternLexer(input)
    val tokens = CommonTokenStream(lexer)
    val parser = PatternParser(tokens)

    val errors = mutableListOf<String>()
    parser.addErrorListener(
        object : BaseErrorListener() {
          override fun syntaxError(
              recognizer: Recognizer<*, *>?,
              offendingSymbol: Any,
              line: Int,
              charPositionInLine: Int,
              msg: String,
              e: RecognitionException?,
          ) {
            errors.add(String.format("line %d:%d, %s", line, charPositionInLine, msg))
          }
        }
    )

    val pattern = parser.pattern()

    if (errors.isNotEmpty()) {
      throw PatternException("Invalid pattern: ${errors.joinToString(", ")}.")
    }

    return PatternVisitor.visitPattern(pattern)
  }

  private object PatternVisitor : PatternParserBaseVisitor<Pattern>() {
    override fun visitPattern(ctx: PatternParser.PatternContext): Pattern {
      if (ctx.cypherPattern() != null) {
        return visitCypherPattern(ctx.cypherPattern())
      }

      return visitSimplePattern(ctx.simplePattern())
    }

    override fun visitCypherPattern(ctx: PatternParser.CypherPatternContext): Pattern {
      if (ctx.cypherRelationshipPattern() != null) {
        val relPattern = ctx.cypherRelationshipPattern()
        val bidirectional =
            ((relPattern.leftArrow() != null && relPattern.rightArrow() != null) ||
                (relPattern.leftArrow() == null && relPattern.rightArrow() == null))
        if (bidirectional) {
          throw PatternException("Direction of relationship pattern must be explicitly set.")
        }

        val keyProperties = mutableSetOf<PropertyMapping>()
        val includeProperties = mutableSetOf<PropertyMapping>()
        val excludeProperties = mutableSetOf<String>()

        if (relPattern.properties() != null && relPattern.properties().propertySelector() != null) {
          extractPropertySelectors(
              relPattern.properties().propertySelector().propSelector(),
              keyProperties,
              includeProperties,
              excludeProperties,
          )
        }

        val startNode =
            enforceKeyProperties(
                visitCypherNodePattern(
                    ctx.cypherNodePattern(if ((relPattern.rightArrow() != null)) 0 else 1)
                )
            )

        val endNode =
            enforceKeyProperties(
                visitCypherNodePattern(
                    ctx.cypherNodePattern(if ((relPattern.rightArrow() != null)) 1 else 0)
                )
            )

        return ensureImplicitWildcard(
            RelationshipPattern(
                LabelOrRelTypeVisitor.visitLabelOrRelType(relPattern.labelOrRelType()),
                enforceExplicitInclusionOnly(startNode),
                enforceExplicitInclusionOnly(endNode),
                includeProperties.contains(PropertyMapping.WILDCARD),
                keyProperties,
                includeProperties.filter { it != PropertyMapping.WILDCARD }.toSet(),
                excludeProperties,
            )
        )
      }

      return ensureImplicitWildcard(
          enforceKeyProperties(visitCypherNodePattern(ctx.cypherNodePattern(0)))
      )
    }

    override fun visitSimplePattern(ctx: PatternParser.SimplePatternContext): Pattern {
      if (ctx.simpleRelationshipPattern() != null) {
        val relPattern = ctx.simpleRelationshipPattern()
        val keyProperties = mutableSetOf<PropertyMapping>()
        val includeProperties = mutableSetOf<PropertyMapping>()
        val excludeProperties = mutableSetOf<String>()

        if (relPattern.properties() != null && relPattern.properties().propertySelector() != null) {
          extractPropertySelectors(
              relPattern.properties().propertySelector().propSelector(),
              keyProperties,
              includeProperties,
              excludeProperties,
          )
        }

        val startNode = enforceKeyProperties(visitSimpleNodePattern(ctx.simpleNodePattern(0)))
        val endNode = enforceKeyProperties(visitSimpleNodePattern(ctx.simpleNodePattern(1)))

        return ensureImplicitWildcard(
            RelationshipPattern(
                SymbolicNameStringVisitor.visitSymbolicNameString(relPattern.symbolicNameString()),
                enforceExplicitInclusionOnly(startNode),
                enforceExplicitInclusionOnly(endNode),
                includeProperties.contains(PropertyMapping.WILDCARD),
                keyProperties,
                includeProperties.filter { it != PropertyMapping.WILDCARD }.toSet(),
                excludeProperties,
            )
        )
      }

      return ensureImplicitWildcard(
          enforceKeyProperties(visitSimpleNodePattern(ctx.simpleNodePattern(0)))
      )
    }

    @Suppress("UNCHECKED_CAST")
    private fun <T : Pattern> ensureImplicitWildcard(pattern: T): T {
      when (pattern) {
        is NodePattern -> {
          if (pattern.includeProperties.isEmpty()) {
            return NodePattern(
                pattern.labels,
                true,
                pattern.keyProperties,
                emptySet(),
                pattern.excludeProperties,
            )
                as T
          }
        }
        is RelationshipPattern -> {
          if (pattern.includeProperties.isEmpty()) {
            return RelationshipPattern(
                pattern.type,
                pattern.start,
                pattern.end,
                true,
                pattern.keyProperties,
                emptySet(),
                pattern.excludeProperties,
            )
                as T
          }
        }
        else ->
            throw IllegalArgumentException("Unsupported pattern type: ${pattern.javaClass.name}.")
      }

      return pattern
    }

    private fun enforceKeyProperties(pattern: NodePattern): NodePattern {
      if (pattern.keyProperties.isEmpty()) {
        throw PatternException("At least one key selector must be specified in node patterns.")
      }

      return pattern
    }

    private fun enforceExplicitInclusionOnly(pattern: NodePattern): NodePattern {
      if (pattern.excludeProperties.isNotEmpty()) {
        throw PatternException(
            "Property exclusions are not allowed on start and end node patterns."
        )
      }

      if (pattern.includeAllValueProperties) {
        throw PatternException(
            "Wildcard property inclusion is not allowed on start and end node patterns."
        )
      }

      return pattern
    }

    override fun visitCypherNodePattern(ctx: PatternParser.CypherNodePatternContext): NodePattern {
      val labels = NodeLabelsVisitor.visitNodeLabels(ctx.nodeLabels())
      return nodePattern(labels, ctx.properties())
    }

    override fun visitSimpleNodePattern(ctx: PatternParser.SimpleNodePatternContext): NodePattern {
      val labels =
          setOf(SymbolicNameStringVisitor.visitSymbolicNameString(ctx.symbolicNameString())) +
              NodeLabelsVisitor.visitNodeLabels(ctx.nodeLabels())
      return nodePattern(labels, ctx.properties())
    }

    private fun nodePattern(
        labels: Set<String>,
        properties: PatternParser.PropertiesContext?,
    ): NodePattern {
      val keyProperties = mutableSetOf<PropertyMapping>()
      val includeProperties = mutableSetOf<PropertyMapping>()
      val excludeProperties = mutableSetOf<String>()

      if (properties?.propertySelector() != null) {
        extractPropertySelectors(
            properties.propertySelector().propSelector(),
            keyProperties,
            includeProperties,
            excludeProperties,
        )
      }

      return NodePattern(
          labels,
          includeProperties.contains(PropertyMapping.WILDCARD),
          keyProperties,
          includeProperties.filter { it != PropertyMapping.WILDCARD }.toSet(),
          excludeProperties,
      )
    }

    private fun extractPropertySelectors(
        selectors: List<PatternParser.PropSelectorContext>,
        keyProperties: MutableSet<PropertyMapping>,
        includeProperties: MutableSet<PropertyMapping>,
        excludeProperties: MutableSet<String>,
    ) {
      selectors.forEach { child ->
        if (child.TIMES() != null) {
          includeProperties.add(PropertyMapping.WILDCARD)
        } else if (child.MINUS() != null) {
          excludeProperties.add(
              PropertyKeyNameVisitor.visitPropertyKeyName(child.propertyKeyName())
          )
        } else if (child.EXCLAMATION() != null) {
          keyProperties += extractPropertyNameOrAlias(child.propertyKeyNameOrAlias())
        } else {
          includeProperties += extractPropertyNameOrAlias(child.propertyKeyNameOrAlias())
        }
      }

      if (includeProperties.isNotEmpty() && excludeProperties.isNotEmpty()) {
        if (!includeProperties.contains(PropertyMapping.WILDCARD)) {
          throw PatternException("Property inclusions and exclusions are mutually exclusive.")
        }
      }
    }
  }

  private fun extractPropertyNameOrAlias(ctx: PropertyKeyNameOrAliasContext): PropertyMapping {
    return if (ctx.propertyKeyName() != null) {
      val property = PropertyKeyNameVisitor.visitPropertyKeyName(ctx.propertyKeyName())
      PropertyMapping(property, property)
    } else {
      PropertyMapping(
          PropertyKeyNameVisitor.visitPropertyKeyName(
              ctx.aliasedPropertyKeyName().propertyKeyName(1)
          ),
          PropertyKeyNameVisitor.visitPropertyKeyName(
              ctx.aliasedPropertyKeyName().propertyKeyName(0)
          ),
      )
    }
  }

  private object NodeLabelsVisitor : PatternParserBaseVisitor<Set<String?>>() {
    override fun visitNodeLabels(ctx: PatternParser.NodeLabelsContext?): Set<String> {
      if (ctx?.labelOrRelType() == null) {
        return emptySet<String>()
      }

      return ctx.labelOrRelType().map { LabelOrRelTypeVisitor.visitLabelOrRelType(it) }.toSet()
    }
  }

  private object LabelOrRelTypeVisitor : PatternParserBaseVisitor<String?>() {
    override fun visitLabelOrRelType(ctx: PatternParser.LabelOrRelTypeContext): String {
      return SymbolicNameStringVisitor.visitSymbolicNameString(ctx.symbolicNameString())
    }
  }

  private object PropertyKeyNameVisitor : PatternParserBaseVisitor<String>() {
    override fun visitPropertyKeyName(ctx: PatternParser.PropertyKeyNameContext): String {
      return SymbolicNameStringVisitor.visitSymbolicNameString(ctx.symbolicNameString())
    }
  }

  private object SymbolicNameStringVisitor : PatternParserBaseVisitor<String>() {
    override fun visitSymbolicNameString(ctx: PatternParser.SymbolicNameStringContext): String {
      return super.visitSymbolicNameString(ctx)
    }

    override fun visitEscapedSymbolicNameString(
        ctx: PatternParser.EscapedSymbolicNameStringContext
    ): String {
      return ctx.text.substring(1, ctx.text.length - 1)
    }

    override fun visitUnescapedSymbolicNameString(
        ctx: PatternParser.UnescapedSymbolicNameStringContext
    ): String {
      return ctx.text
    }
  }
}
